includes/Modules/Analytics/Shirshu_Class.php

```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ * Class Google\Site_Kit\Modules\Analytics\Shirshu_Class
+ *
+ * @package   Google\Site_Kit\Modules\Analytics
+ * @copyright 2019 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Modules\Analytics;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Modules\Analytics\Plugin_Detector;
+use Google\Site_Kit\Modules\Analytics\Measurement_Code_Injector;
+
+/**
+ * Main Shirshu_Class class
+ *
+ * @class Shirshu_Class
+ */
+final class Shirshu_Class {
+
+	/**
+	 * List of plugins ShirshuClass supports for event tracking
+	 *
+	 * @var array of strings
+	 */
+	private $supported_plugins;
+
+	/**
+	 * Single instance of the class
+	 *
+	 * @var Shirshu_Class
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Instance of the plugin detector
+	 *
+	 * @var Plugin_Detector
+	 */
+	protected $plugin_detector = null;
+
+	/**
+	 * Instance of the event injector
+	 *
+	 * @var Measurement_Code_Injector
+	 */
+	protected $measurement_code_injector = null;
+
+	/**
+	 * Plugin context.
+	 *
+	 * @since 1.0.0
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * Returns the main Shirshu_Class instance
+	 *
+	 * @param Context $module_context used to check for AMP.
+	 * @return Shirshu_Class - Main instance
+	 */
+	public static function get_instance( $module_context ) {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self( $module_context );
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Shirshu_Class constructor
+	 *
+	 * @param Context $module_context context used to check for AMP.
+	 */
+	private function __construct( $module_context ) {
+		$this->context           = $module_context;
+		$this->supported_plugins = array(
+			'Contact Form 7'   => 'WPCF7_PLUGIN_DIR',
+			'Formidable Forms' => 'load_formidable_forms',
+			'Ninja Forms'      => 'NF_PLUGIN_DIR',
+			'WooCommerce'      => 'WC_PLUGIN_FILE',
+			'WPForms'          => 'WPFORMS_PLUGIN_DIR',
+			'WPForms Lite'     => 'WPFORMS_PLUGIN_DIR',
+		);
+		$this->plugin_detector   = new Plugin_Detector( $this->supported_plugins );
+		$this->get_active_plugins();
+	}
```

Comment:
Revisiting. this class, I'm starting to think we should remove it entirely. All it does is detect the active plugins via the `Plugin_Detector` class. While I'm all for separating concerns, I think this is a bit too much here, especially since the list of supported plugins is actually tied to which `Measurement_Event_List` derived classes we implement.

I'd suggest to move this list of supported plugins and usage of `Plugin_Detector` into the `Measurement_Code_Injector` class and then remove this class. With this change, `Measurement_Code_Injector` would then become the new entry point for the advanced tracking functionality.

Reply:
I can see where you're coming from. The original idea was to have a driver class that uses the functionality of the Plugin Detector and Code Injector, since those two classes perform separate tasks. However, the flow of things got messed up when we had to incorporate wordpress hooks. Knowing that this class won't be instantiated until after the plugins are loaded, could I try rearranging some things to make it closer to our original idea? If it still seems like overkill, we can definitely remove the class entirely.

Reply:
Sure I'm open for alternatives, however one thing that should be avoided is running this logic to detect the active supported plugins when not needed. In other words, this probably needs to become dependent on the `Measurement_Code_Injector` class in some way, since that class has awareness of whether event tracking code should be injected or not (if Site Kit did not enqueue `google_gtagjs`, there is no point in running any advanced tracking logic).

Reply:
Okay I will keep that in mind.